source/ocproute.go

```diff
@@ -0,0 +1,281 @@
+	versioned "github.com/openshift/client-go/route/clientset/versioned"
+	extInformers "github.com/openshift/client-go/route/informers/externalversions"
+	routeInformer "github.com/openshift/client-go/route/informers/externalversions/route/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+// ocpRouteSource is an implementation of Source for OpenShift Route objects.
+// Route implementation will use the spec.host value for the hostname
+// Use targetAnnotationKey to explicitly set Endpoint. (useful if the router
+// does not update, or to override with alternative endpoint)
+type ocpRouteSource struct {
+	client                   versioned.Interface
+	namespace 	             string
+	annotationFilter         string
+	fqdnTemplate             *template.Template
+	combineFQDNAnnotation    bool
+	ignoreHostnameAnnotation bool
+	routeInformer            routeInformer.RouteInformer
+}
+
+// NewOcpRouteSource creates a new ocpRouteSource with the given config.
+func NewOcpRouteSource(
+	ocpClient versioned.Interface,
+	namespace string,
+	annotationFilter string,
+	fqdnTemplate string,
+	combineFQDNAnnotation bool,
+	ignoreHostnameAnnotation bool,
+) (Source, error) {
+	var (
+		tmpl *template.Template
+		err  error
+	)
+	if fqdnTemplate != "" {
+		tmpl, err = template.New("endpoint").Funcs(template.FuncMap{
+			"trimPrefix": strings.TrimPrefix,
+		}).Parse(fqdnTemplate)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Use shared informer to listen for add/update/delete of Routes in the specified namespace.
+	// Set resync period to 0, to prevent processing when nothing has changed.
+	informerFactory := extInformers.NewFilteredSharedInformerFactory(ocpClient, 0, namespace, nil)
+	routeInformer := informerFactory.Route().V1().Routes()
+
+	// Add default resource event handlers to properly initialize informer.
+	routeInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) {
+			},
+		},
+	)
+
+	// TODO informer is not explicitly stopped since controller is not passing in its channel.
+	informerFactory.Start(wait.NeverStop)
+
+	// wait for the local cache to be populated.
+	err = wait.Poll(time.Second, 60*time.Second, func() (bool, error) {
+		return routeInformer.Informer().HasSynced(), nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to sync cache: %v", err)
+	}
+
+	return &ocpRouteSource{
+		client:                   ocpClient,
+		namespace:                namespace,
+		annotationFilter:         annotationFilter,
+		fqdnTemplate:             tmpl,
+		combineFQDNAnnotation:    combineFQDNAnnotation,
+		ignoreHostnameAnnotation: ignoreHostnameAnnotation,
+		routeInformer:            routeInformer,
+	}, nil
+}
+
+func (ors *ocpRouteSource) AddEventHandler(handler func() error, stopChan <-chan struct{}, minInterval time.Duration) {
```

Comment:
Why is this function empty? At a minimum there should be a TODO comment added here.

Reply:
Since every Source type needs an AddEventHandler function I just copied this one from ingressroute.go.
Added a TODO.